File: 108-convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {

private:
    TreeNode* convert(vector<int>& arr, int low, int high) {
        if (low > high) {
            return nullptr;
        }

        int mid = (low + high) / 2;

        TreeNode* node = new TreeNode(arr[mid]);

        

        node->left = convert(arr, low, mid - 1);
        node->right = convert(arr, mid + 1, high);

        return node;
    }

public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if (sizeof(nums) == 0) {
        return nullptr;
        }

        return convert(nums, 0, nums.size() - 1);
    }
};

File: 108-convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def addNode(arr, low, high):
            if low > high:
                return None

            mid = (low + high) // 2
            node = TreeNode(val=arr[mid])

            node.left = addNode(arr, low, mid-1)
            node.right = addNode(arr, mid+1, high)

            return node


        if len(nums) == 0:
            return None

        low , high = 0, len(nums) -1

        head = addNode(nums, low, high)


        return head


        

        